extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/VertxSecurityIdentityAssociation.java

```diff
@@ -0,0 +1,78 @@
+package io.quarkus.vertx.http.runtime.security;
+
+import static io.quarkus.vertx.http.runtime.security.QuarkusHttpUser.DEFERRED_IDENTITY_KEY;
+
+import java.security.Principal;
+
+import jakarta.enterprise.context.RequestScoped;
+import jakarta.enterprise.inject.Produces;
+
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.spi.runtime.AbstractSecurityIdentityAssociation;
+import io.smallrye.common.vertx.VertxContext;
+import io.smallrye.mutiny.Uni;
+import io.vertx.core.Context;
+import io.vertx.core.Vertx;
+import io.vertx.ext.web.RoutingContext;
+
+@RequestScoped
+public class VertxSecurityIdentityAssociation extends AbstractSecurityIdentityAssociation {
+
+    @Produces
+    @RequestScoped
+    public Principal principal() {
+        return new Principal() {
+            @Override
+            public String getName() {
+                return getIdentity().getPrincipal().getName();
+            }
+        };
+    }
+
+    @Override
+    public Uni<SecurityIdentity> getDeferredIdentity() {
+        RoutingContext routingContext = getRoutingContext();
+        if (routingContext != null) {
+            SecurityIdentity securityIdentity = getSecurityIdentityFromCtx(routingContext);
+            if (securityIdentity != null) {
+                return Uni.createFrom().item(securityIdentity);
+            }
+            Uni<SecurityIdentity> identityUni = routingContext.get(DEFERRED_IDENTITY_KEY);
+            if (identityUni != null) {
+                return identityUni;
+            }
+        }
+        return super.getDeferredIdentity();
+    }
+
+    @Override
+    public SecurityIdentity getIdentity() {
+        RoutingContext routingContext = getRoutingContext();
+        if (routingContext != null) {
+            SecurityIdentity securityIdentity = getSecurityIdentityFromCtx(routingContext);
+            if (securityIdentity != null) {
+                return securityIdentity;
+            }
+            securityIdentity = QuarkusHttpUser.getSecurityIdentityBlocking(routingContext, null);
+            if (securityIdentity != null) {
+                return securityIdentity;
+            }
+        }
+        return super.getIdentity();
+    }
+
+    private static SecurityIdentity getSecurityIdentityFromCtx(RoutingContext routingContext) {
+        if (routingContext.user() instanceof QuarkusHttpUser quarkusHttpUser) {
+            return quarkusHttpUser.getSecurityIdentity();
+        }
+        return null;
+    }
+
+    private static RoutingContext getRoutingContext() {
+        Context context = Vertx.currentContext();
+        if (context != null && VertxContext.isDuplicatedContext(context)) {
+            return context.getLocal(HttpSecurityUtils.ROUTING_CONTEXT_ATTRIBUTE);
```

Comment:
Could you use `io.smallrye.common.vertx.ContextLocals`? It would simplify the switch to Vert.x 5.

Reply:
Done, thanks